test/functional/kernel/hyperplane/test-kernel-hyperplane-3D.cpp.in

```diff
@@ -0,0 +1,228 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-23, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+//
+// test/include headers
+//
+#include "RAJA_test-base.hpp"
+#include "RAJA_test-camp.hpp"
+#include "RAJA_test-index-types.hpp"
+#include "RAJA_test-kernel-tile-size.hpp"
+
+#include "RAJA_test-reducepol.hpp"
+
+// for data types
+#include "RAJA_test-reduce-types.hpp"
+#include "RAJA_test-forall-data.hpp"
+
+//
+// Header for tests in ./tests directory
+//
+// Note: CMake adds ./tests as an include dir for these tests.
+//
+#include "test-kernel-hyperplane-@TEST_TYPE@.hpp"
+
+
+//
+// Exec pols for kernel tile tests
+//
+
+using SequentialKernelHyperplaneExecPols =
+  camp::list<
+
+    RAJA::KernelPolicy<
+      RAJA::statement::For<0, RAJA::seq_exec,
+        RAJA::statement::Hyperplane<1, RAJA::seq_exec, RAJA::ArgList<2, 3>, RAJA::seq_exec,
+          RAJA::statement::Lambda<0>
+        >
+      >
+    >
+
+    // Sequential tiling with Hyperplane is not yet implemented
+    //RAJA::KernelPolicy<
+    //  RAJA::statement::For<0, RAJA::seq_exec,
+    //    RAJA::statement::Tile<2, RAJA::tile_fixed<11>, RAJA::seq_exec,
+    //      RAJA::statement::Tile<3, RAJA::tile_fixed<13>, RAJA::seq_exec,
+    //        RAJA::statement::For<2, RAJA::seq_exec,
+    //          RAJA::statement::For<3, RAJA::seq_exec,
+    //            RAJA::statement::Hyperplane<1, RAJA::seq_exec, RAJA::ArgList<2, 3>, RAJA::seq_exec,
+    //              RAJA::statement::Lambda<0>
+    //            >
+    //          >
+    //        >
+    //      >
+    //    >
+    //  >
+    //>
```

Comment:
This is also something I'd like people to weigh in on whether it should be an issue.

The problem lies in that the standard `KernelPolicy` will auto-collapse a series of nested `For`s. When the compiler reaches `Hyperplane`, all of the arguments have already been consumed by the previously collapsed `For` and `Tile` statements, causing RAJA to fail on a static assert. This is why the policy above this one works; all of the `Hyperplane` arguments there are not consumed by other statements.

However, `CudaKernel` and `HipKernel` can operate on this test because those implementations delay consumption of the range arguments until the innermost level (`Hyperplane`), and begins collapsing (if possible) at the latest possible moment. This is why the `CudaKernel` and `HipKernel` variations of this test can seemingly reuse ranges from previous statements.

If I've understood these implementations correctly, to implement this same interface for our standard `KernelPolicy`, we'd need to add quite a bit more complexity to our `sequential` kernel policy. I'm not sure this is a good idea, but someone can please correct me if I've misunderstood how kernel works. There is also no demand for a purely sequential tiled Hyperplane, and until someone wants it I don't think we need to implement it.

Reply:
It sounds like capturing this knowledge in an issue is a good idea, then we can make changes if needed later.

Reply:
@rhornung67 Shall we just punt on this until there is actual demand for this use case? I've documented it here https://github.com/LLNL/RAJA/issues/1493, so if others are interested we can continue discussion in the issue.